src/LiquidityPool.sol

```diff
@@ -70,6 +68,24 @@ contract LiquidityPool is Auth {
         address indexed sender, address indexed receiver, address indexed owner, uint256 assets, uint256 shares
     );
 
+    /// @dev function either called by a ward or message.sender has approval to spent sender´s tokens
+    modifier withTokenApproval(address sender, uint256 amount) {
+        require(
+            wards[msg.sender] == 1 || msg.sender == sender || share.allowance(sender, msg.sender) >= amount,
+            "LiquidityPool/no-token-allowance"
+        );
+        _;
+    }
+
+    /// @dev function either called by a ward or message.sender has approval to spent sender´s currency
+    modifier withCurrencyApproval(address sender, uint256 amount) {
+        require(
+            wards[msg.sender] == 1 || msg.sender == sender || ERC20Like(asset).allowance(sender, msg.sender) >= amount,
+            "LiquidityPool/no-currency-allowance"
+        );
+        _;
+    }
+
     constructor() {
```

Comment:
Unrelate to this PR, but why aren't `poolId` and `trancheId` immutable and passed in the constructor and is `setPoolDetails` used instead?